polars/polars-arrow/src/kernels/ewm/average.rs

```diff
@@ -1,47 +1,44 @@
+use std::ops::AddAssign;
+
 use arrow::array::PrimitiveArray;
 use arrow::types::NativeType;
 use num::Float;
 
 use crate::utils::CustomIterTools;
 
-// See:
-// https://github.com/pola-rs/polars/issues/2148
-// https://stackoverflow.com/a/51392341/6717054
-
 pub fn ewm_mean<T>(
     xs: &PrimitiveArray<T>,
     alpha: T,
     adjust: bool,
     min_periods: usize,
 ) -> PrimitiveArray<T>
 where
-    T: Float + NativeType,
+    T: Float + NativeType + AddAssign,
 {
-    let mut denom = T::zero();
     let one_sub_alpha = T::one() - alpha;
+
+    let mut opt_mean = None;
     let mut non_null_cnt = 0usize;
 
-    let mut opt_ewma = None;
+    let wgt = alpha;
+    let mut wgt_sum = if adjust { T::zero() } else { T::one() };
```

Comment:
Nice that we could store the adjust info in a `value` that we use in arithmetic.

There is an opportunity for future optimization where we have two separate loops depending on the `ajust` parameter.

I do think we need to check `alpha` in combination with `adjust` as this could lead to divide by zero on line 35.

Reply:
> There is an opportunity for future optimization where we have two separate loops depending on the adjust parameter.

Yes, definitely. I've optimized the code for simplicity here, not necessarily performance. I think it is important to implement all the desired functionality as simply (and correctly) as possible before heavily optimizing. For example, we may still want to include support for the `ignore_na` or `axis` parameter.

Ah, thanks for reminding me about `alpha`, I did not think carefully about the edge cases of zero and one. Here is the output of this MR

```python
In [2]: pl.Series([1., 2., 3.]).ewm_mean(alpha=0.0, adjust=True)
Out[2]: 
shape: (3,)
Series: '' [f64]
[
	NaN
	NaN
	NaN
]
# NOTE: master outputs *incorrect* values

In [3]: pl.Series([1., 2., 3.]).ewm_mean(alpha=0.0, adjust=False)
Out[3]: 
shape: (3,)
Series: '' [f64]
[
	1.0
	1.0
	1.0
]
# NOTE: master outputs *correct* values

In [4]: pl.Series([1., 2., 3.]).ewm_mean(alpha=1.0, adjust=True)
Out[4]: 
shape: (3,)
Series: '' [f64]
[
	1.0
	2.0
	3.0
]
# NOTE: master outputs *correct* values

In [5]: pl.Series([1., 2., 3.]).ewm_mean(alpha=1.0, adjust=False)
Out[5]: 
shape: (3,)
Series: '' [f64]
[
	1.0
	2.0
	3.0
]
# NOTE: master outputs *correct* values
```

And this is how `pandas` behaves

```python
In [17]: pd.Series([1., 2., 3.]).ewm(alpha=0.0, adjust=True).var()
ValueError: alpha must satisfy: 0 < alpha <= 1

In [17]: pd.Series([1., 2., 3.]).ewm(alpha=0.0, adjust=False).var()
ValueError: alpha must satisfy: 0 < alpha <= 1

In [9]: pd.Series([1., 2., 3.]).ewm(alpha=1.0, adjust=True).mean()
Out[9]: 
0    1.0
1    2.0
2    3.0
dtype: float64

In [10]: pd.Series([1., 2., 3.]).ewm(alpha=1.0, adjust=False).mean()
Out[10]: 
0    1.0
1    2.0
2    3.0
dtype: float64
```

One option would be to disallow zero and/or one -- and there's reasonable justification for that.

- `alpha=0` and `adjust=True`: Mathematically, this is just a simple mean
- `alpha=0` and `adjust=False`: The math says to just forward fill the first value
- `alpha=1` and `adjust=True`: This is a no-op, as we give zero weight to previous observations
-  `alpha=1` and `adjust=False`: This is a no-op, as we give zero weight to previous observations

Alternatively, I could just dispatch `alpha=0` and `adjust=True` to compute a simple mean... any particular preference?

Reply:
> Alternatively, I could just dispatch `alpha=0` and `adjust=True` to compute a simple mean... any particular preference?

Dispatching to a simple mean computation won't work. For starters, I don't see expanding mean functionality implemented anywhere. But even if we did have it, it wouldn't work for the variance and standard deviation calculations. I've followed `pandas` lead and required `0 < alpha <= 1` in my latest commit. Our docstrings on master actually specify `0 < alpha < 1`.

If this approach is acceptable, I will then open a follow-on PR implementing a separate `expanding_mean` function.